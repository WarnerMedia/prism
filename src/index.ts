// Copyright (c) Warner Media, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
import { Psm, PsmConfig } from "./Psm";

const psm = new Psm();

export * from "./Psm";
declare global {
  interface Window {
    psmMgrConfig: PsmConfig;
    psmMgr: {
      psm_test_ts: boolean;
      psmEnvironment: string;
      psmSendGeoInit: boolean;
      psmHost: string;
      psmGeoActualCookieValue: string;
    };
    psmCore?: { psm?: { browser?: string } };
  }
  interface Crypto {
    randomUUID: () => string;
  }
}

window.psmMgr = {
  ...window.psmMgr,
  psm_test_ts: true, // FIXME: for dev purposes, allows our logger function to log to the console
};
