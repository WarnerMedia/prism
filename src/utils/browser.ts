/* eslint-disable @typescript-eslint/no-explicit-any */

const win = window as any;
const nav = navigator as any;
const doc = document as any;

/**
 * `navigator.cookieEnabled` cannot detect custom or nuanced cookie-blocking configurations.
 * For example, when blocking cookies via the Advanced Privacy Settings in IE, it always returns true.
 * There can be issues with site-specific exceptions.
 *
 * @see https://github.com/Modernizr/Modernizr/blob/master/feature-detects/cookies.js Taken from here
 *  
 * The MIT License (MIT)
 * 
 * Copyright (c) 2021 The Modernizr Team | Modernizr 4.0.0-alpha
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
export function cookiesEnabled(): boolean {
  // try..catch because some in situations `document.cookie` is exposed but throws a
  // SecurityError if you try to access it; e.g. documents created from data URIs
  // or in sandboxed iframes (depending on flags/context)
  try {
    // Create test cookie
    doc.cookie = "cookietest=1; SameSite=Strict;";
    const result = doc.cookie.indexOf("cookietest=") !== 1;
    // Delete test cookie
    doc.cookie =
      "cookietest=1; SameSite=Strict; expires=Thu, 01-Jan-1970 00:00:01 GMT";
    return result;
  } catch (err) {
    return false;
  }
}

/**
 * Checks that localStorage is available, accounting for
 * a `SecurityError` being thrown in Firefox if "always ask" is enabled
 * for cookies.
 */
export function hasLocalStorage(): boolean {
  try {
    return !!win.localStorage;
  } catch (err) {
    return true;
  }
}

/**
 * Checks that localStorage is accessible, accounting for private mode in some
 * browsers.
 */
export function localStorageAccessible(): boolean {
  const test = "__psm_test__";

  if (!hasLocalStorage()) {
    return false;
  }
  try {
    win.localStorage.setItem(test, test);
    win.localStorage.removeItem(test);
    return true;
  } catch (err) {
    return false;
  }
}

/**
 * Gets the current viewport.
 *
 * Code based on:
 * - http://andylangton.co.uk/articles/javascript/get-viewport-size-javascript/
 * - http://responsejs.com/labs/dimensions/
 */
export function detectViewport(): string {
  let alias: Window | HTMLElement = win;
  let modifier = "inner";

  if (!("innerWidth" in win)) {
    modifier = "client";
    alias = document.documentElement || document.body;
  }

  const width = alias[modifier + "Width"];
  const height = alias[modifier + "Height"];

  if (width >= 0 && height >= 0) {
    return width + "x" + height;
  } else {
    return null;
  }
}

/**
 * Gets the total width and height of the current document
 */
export function detectDocumentSize(): number[] {
  const de = document.documentElement;
  const be = document.body;
  const bh = be ? Math.max(be.offsetHeight, be.scrollHeight) : 0;
  const w = Math.max(de.clientWidth, de.offsetWidth, de.scrollWidth);
  const h = Math.max(de.clientHeight, de.offsetHeight, de.scrollHeight, bh);

  return isNaN(w) || isNaN(h) ? [0, 0] : [w, h];
}

export function getScreenResolution(): string {
  return win.screen.width + "x" + win.screen.height;
}

/**
 * Extract hostname from URL
 */
export function getHostname(url: string): string {
  const matches = new RegExp("^(?:(?:https?|ftp):)/*(?:[^@]+@)?([^:/#]+)").exec(
    url
  );

  return matches ? matches[1] : url;
}

/**
 * Get page referrer. In the case of a single-page app,
 * if the URL changes without the page reloading, pass
 * in the old URL. It will be returned unless overriden
 * by a "refer(r)er" parameter in the querystring.
 *
 * @param string oldLocation Optional.
 * @return string The referrer
 */
export function getReferrer(oldLocation?: string): string {
  let referrer = "";

  if (oldLocation) {
    return oldLocation;
  }

  try {
    referrer = win.top.document.referrer;
  } catch (err) {
    if (win.parent) {
      try {
        referrer = win.parent.document.referrer;
      } catch (err) {
        referrer = "";
      }
    }
  }

  if (referrer === "") {
    referrer = document.referrer;
  }

  return referrer;
}

/**
 * Throttles a function, i.e. do not allow the function to be invoked more than
 * once every X milliseconds.
 *
 * @param callback The function to throttle
 * @param wait The amount of time in milliseconds by which to throttle
 */
export function throttle<T extends []>(
  callback: (..._: T) => void,
  wait: number
): (..._: T) => void {
  let queuedToRun: number | undefined;
  let previouslyRun: number;

  return function invokeFn(...args: T) {
    const now = Date.now();
    queuedToRun = clearTimeout(queuedToRun) as undefined;
    if (!previouslyRun || now - previouslyRun >= wait) {
      callback(...args);
      previouslyRun = now;
    } else {
      queuedToRun = window.setTimeout(
        invokeFn.bind(null, ...args),
        wait - (now - previouslyRun)
      );
    }
  };
}

/**
 * Checks that a DOM element is not hidden by another element with a higher z-index
 * using document.getElementFromPoint(). To be safe, the center of the expected element is used
 * to account for border radius.
 */
export function isElementVisible(el: Element) {
  const rect = el.getBoundingClientRect();

  return el.contains(
    doc.elementFromPoint(
      rect.right - rect.width / 2,
      rect.bottom - rect.height / 2
    )
  );
}

/**
 * Determines the current browser
 */
export function setBrowser(): string {
  const userAgent = navigator.userAgent;

  const isSafari =
    userAgent.indexOf("Safari") > -1 &&
    userAgent.indexOf("Chrome") === -1 &&
    userAgent.indexOf("Chromium") === -1 &&
    userAgent.indexOf("CriOS") === -1 &&
    userAgent.indexOf("EdgiOS") === -1 &&
    userAgent.indexOf("FxiOS") === -1;

  const isFirefox =
    (userAgent.indexOf("Firefox") > -1 || userAgent.indexOf("FxiOS") > -1) &&
    userAgent.indexOf("Seamonkey") === -1;

  const isChrome =
    (userAgent.indexOf("Chrome") > -1 || userAgent.indexOf("CriOS") > -1) &&
    userAgent.indexOf("Chromium") === -1 &&
    userAgent.indexOf("Edg") === -1 &&
    userAgent.indexOf("OPR") === -1;

  const isEdge = userAgent.indexOf("Edg") > -1;

  switch (true) {
    case isSafari:
      return "safari";
    case isFirefox:
      return "firefox";
    case isChrome:
      return "chrome";
    case isEdge:
      return "edge";
    default:
      return "unknown";
  }
}

/**
 * Return a value from a name-value pair in a query string
 */
export function fromQueryString(field: string, url: string) {
  const match = new RegExp("^[^#]*[?&]" + field + "=([^&#]*)").exec(url);
  if (!match) {
    return null;
  }
  return decodeURIComponent(match[1].replace(/\+/g, " "));
}

/**
 * Add name-value pair to the query string of a URL
 */
export function updateQueryString(url: string, name: string, value: string) {
  const initialQsParams = name + "=" + value;
  const hashSplit = url.split("#");
  const qsSplit = hashSplit[0].split("?");
  const beforeQuerystring = qsSplit.shift();
  // Necessary because a querystring may contain multiple question marks
  let querystring = qsSplit.join("?");
  if (!querystring) {
    querystring = initialQsParams;
  } else {
    // Whether this is the first time the link has been decorated
    let initialDecoration = true;
    const qsFields = querystring.split("&");
    for (let i = 0; i < qsFields.length; i++) {
      if (qsFields[i].substr(0, name.length + 1) === name + "=") {
        initialDecoration = false;
        qsFields[i] = initialQsParams;
        querystring = qsFields.join("&");
        break;
      }
    }
    if (initialDecoration) {
      querystring = initialQsParams + "&" + querystring;
    }
  }
  hashSplit[0] = beforeQuerystring + "?" + querystring;
  return hashSplit.join("#");
}
