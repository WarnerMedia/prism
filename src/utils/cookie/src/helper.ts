import { CookieOptions } from ".";

export const debug = (...args: (string | object | number | boolean)[]) => {
  if (window.location.search.search(/[?&]psm_debug=[1t]/) !== -1) {
    console.info(...args);
  }
};

export function cookieHelper(
  name?: string,
  value?: string,
  options?: Partial<CookieOptions>
): string | object | void {
  switch (arguments.length) {
    case 3:
    case 2:
      return set(name, value, options);
    case 1:
      return get(name);
    default:
      return all();
  }
}

/**
 * Set cookie `name` to `value`.
 */
function set(name: string, value: string, options: Partial<CookieOptions>) {
  options = options || {};
  let str = encode(name) + "=" + encode(value);

  if (options.encode === false) {
    str = name + "=" + value;
  }

  if (null == value) options.maxage = -1;

  if (options.maxage && !options.expires) {
    options.expires = new Date(+new Date() + options.maxage);
  }

  if (options.path) str += "; Path=" + options.path;
  if (options.domain) str += "; Domain=" + options.domain;
  if (options.expires) str += "; Expires=" + options.expires.toUTCString();
  if (options.samesite) str += "; SameSite=" + options.samesite;
  if (options.secure) str += "; Secure";

  debug("[COOKIE]: cookie.helper.set()", str, options);

  document.cookie = str;
}

/**
 * Return all cookies.
 */
function all(): object {
  let str: string;
  try {
    str = document.cookie;
  } catch (err) {
    if (typeof console !== "undefined" && typeof console.error === "function") {
      console.error(err.stack || err);
    }
    return {};
  }
  return parse(str);
}

/**
 * Get cookie `name`.
 */
function get(name: string): string {
  const val = all()[name];

  debug(`[COOKIE]: helper.get() - getting ${name} cookie value ${val}`);

  return val;
}

/**
 * Parse cookie `str`.
 */
function parse(str: string): { [key: string]: string } {
  const obj = {};
  const pairs = str.split(/ *; */);
  let pair: string[];

  if ("" == pairs[0]) return obj;

  for (let i = 0; i < pairs.length; ++i) {
    const cookieName = pairs[i].substr(0, pairs[i].indexOf("="));
    const cookieValue = pairs[i].substr(pairs[i].indexOf("=") + 1);
    obj[decode(cookieName)] = decode(cookieValue);
  }

  return obj;
}

function encode(value: string | number | boolean): string {
  try {
    return encodeURIComponent(value);
  } catch (err) {
    // ignore
  }
}

function decode(value: string): string {
  try {
    return decodeURIComponent(value);
  } catch (err) {
    // ignore
  }
}
