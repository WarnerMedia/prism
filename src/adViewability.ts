// Copyright (c) Warner Media, LLC. All rights reserved. Licensed under the MIT license.
// See the LICENSE file for license information.
import { isElementVisible } from "./utils/browser";

declare global {
  interface Window {
    googletag: googletag.Googletag;
  }
}

export interface SlotMetrics {
  slotId: string;
  slotSize: string;
  adUnitPath?: string;
  totalViewTime: number;
  lastViewStarted: number;
  visibleOnStart: boolean;
  adWasViewed: boolean;
  rfv?: string;
  creativeId?: string;
  lineItemId?: string;
  totalHoverTime: number;
  lastHoverStarted: number;
}

const win = window;
const doc = document;
let adObserver: IntersectionObserver;

export const registeredSlots: SlotMetrics[] = [];

win.googletag =
  win.googletag || ({ cmd: [] } as unknown as googletag.Googletag);

const log = (...args: string[]) => {
  if (win.location.search.search(/[?&]psm_debug=[1t]/) !== -1) {
    // console.log('[PSM]:', ...args);
  }
};

const timeInSeconds = () => Math.round(Date.now() / 1000);

const addDistinctRegisteredSlot = (slot: SlotMetrics) => {
  if (!registeredSlots.some((regSlot) => slot.slotId === regSlot.slotId)) {
    registeredSlots.push(slot);
  }
};

export function initAdViewListeners(): void {
  adObserver = new IntersectionObserver(
    (entries) => {
      entries.forEach((entry) => {
        const slotId = entry.target.id;
        const inViewPercentage = Math.round(entry.intersectionRatio) * 100;
        const rect = entry.boundingClientRect;

        if (entry.isIntersecting) {
          if (
            entry.intersectionRatio >= 0.5 &&
            isElementVisible(entry.target)
          ) {
            log(
              `slotVisibilityChanged - ${slotId} is in view. (${inViewPercentage}% in view)`
            );
            registeredSlots.forEach((slot) => {
              if (slot.slotId === slotId) {
                slot.visibleOnStart = true;
                slot.adWasViewed = true;
                slot.slotSize = `${rect.width}x${rect.height}`;
                updateSlotTimer(slot);
              }
            });
          }
        } else {
          // if an ad is not considered in view, flag it as not visible and reset lastViewStarted
          log(
            `slotVisibilityChanged - ${slotId} is no longer in view. (${inViewPercentage}% in view)`
          );
          registeredSlots.forEach((slot) => {
            if (slot.slotId === slotId) {
              slot.visibleOnStart = false;
              slot.lastViewStarted = 0;
              slot.lastHoverStarted = 0;
            }
          });
        }
      });
    },
    {
      root: null,
      rootMargin: "0px",
      threshold: [0.0, 0.5, 1.0],
    }
  );

  win.googletag.cmd.push(() => {
    win.googletag
      .pubads()
      .addEventListener(
        "slotRenderEnded",
        (e: googletag.events.SlotRenderEndedEvent) => {
          const slotId = e.slot.getSlotElementId();
          const adUnitPath = e.slot.getAdUnitPath();
          const elem = doc.getElementById(slotId);
          let slotSize = "0x0";

          if (e.size) {
            if (typeof e.size === "string") {
              slotSize = e.size;
            } else {
              slotSize = e.size.join("x");
            }
          }

          const slot: SlotMetrics = {
            slotId,
            slotSize,
            adUnitPath,
            totalViewTime: 0,
            lastViewStarted: 0,
            visibleOnStart: false,
            adWasViewed: false,
            rfv: e.slot.getTargeting("rfv")[0],
            creativeId: e.slot.getTargeting("creativeId")[0],
            lineItemId: e.slot.getTargeting("lineItemId")[0],
            totalHoverTime: 0,
            lastHoverStarted: 0,
          };

          addDistinctRegisteredSlot(slot);

          if (elem) {
            adObserver.observe(elem);
            adListeners(elem, slot);
          }
        }
      );

    win.googletag
      .pubads()
      .addEventListener(
        "slotVisibilityChanged",
        (e: googletag.events.SlotVisibilityChangedEvent) => {
          const slotId = e.slot.getSlotElementId();
          registeredSlots.forEach((slot) => {
            if (slot.slotId === slotId) {
              slot.rfv = e.slot.getTargeting("rfv")[0];
              (slot.creativeId = e.slot.getTargeting("creativeId")[0]),
                (slot.lineItemId = e.slot.getTargeting("lineItemId")[0]);
            }
          });
        }
      );

    win.googletag
      .pubads()
      .getSlots()
      .forEach((slot: googletag.Slot) => {
        const slotId = slot.getSlotElementId();
        const elem = doc.getElementById(slotId);
        let rect = {
          width: 0,
          height: 0,
        };

        if (elem) {
          rect = elem.getBoundingClientRect();
        }

        const adSlot: SlotMetrics = {
          slotId,
          slotSize: rect
            ? `${Math.round(rect.width)}x${Math.round(rect.height)}`
            : "0x0",
          adUnitPath: slot.getAdUnitPath(),
          totalViewTime: 0,
          lastViewStarted: 0,
          visibleOnStart: false,
          adWasViewed: false,
          rfv: slot.getTargeting("rfv")[0],
          creativeId: slot.getTargeting("creativeId")[0],
          lineItemId: slot.getTargeting("lineItemId")[0],
          totalHoverTime: 0,
          lastHoverStarted: 0,
        };

        addDistinctRegisteredSlot(adSlot);

        if (elem) {
          adObserver.observe(elem);
          adListeners(elem, adSlot);
        }
      });
  });

  // listen for the inbrain.loaded event and retrieve array of DOM IDs
  // each Inbrain promo item should be registered as a distinct ad slot
  doc.addEventListener("inbrain.loaded", ((e: CustomEvent) => {
    e.detail.forEach((target) => {
      const elem = doc.getElementById(target);

      if (elem) {
        const rect = elem.getBoundingClientRect();

        addDistinctRegisteredSlot({
          slotId: target,
          slotSize: rect
            ? `${Math.round(rect.width)}x${Math.round(rect.height)}`
            : "0x0",
          totalViewTime: 0,
          lastViewStarted: 0,
          visibleOnStart: false,
          adWasViewed: false,
          totalHoverTime: 0,
          lastHoverStarted: 0,
        });

        adObserver.observe(elem);
      }
    });
  }) as EventListener);
}

export function onHiddenAds() {
  // remove ad slots from the set of visible slots and mark them as inactive
  // so we can restore them when a user tabs back into the document
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      slot.lastViewStarted = 0;
    }
    const elem = doc.getElementById(slot.slotId);

    if (elem) {
      adObserver.unobserve(elem);
    }
  });
}

export function onVisibleAds() {
  // reset the last view started to the current time when the document first becomes visible
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      slot.lastViewStarted = timeInSeconds();
    }
    const elem = doc.getElementById(slot.slotId);

    if (elem) {
      adObserver.observe(elem);
    }
  });
}

export function setAdViewMetrics() {
  registeredSlots.forEach((slot) => {
    if (slot.visibleOnStart) {
      updateSlotTimer(slot);
    }
  });
}

export function resetAdViewMetrics() {
  registeredSlots.forEach((slot) => {
    slot.totalViewTime = 0;
    slot.totalHoverTime = 0;
    slot.adWasViewed = false;
  });
}

function updateSlotTimer(slot: SlotMetrics) {
  const lastStarted = slot.lastViewStarted;
  const currentTime = timeInSeconds();

  // if the timer is currently running
  if (lastStarted) {
    // compute the difference between the current time and the start time
    // to determine the number of seconds the ad slot has been visible since the last
    // time it became visible
    const diff = currentTime - lastStarted;

    slot.totalViewTime += diff;
  }

  // update the last viewed time to the current time
  slot.lastViewStarted = currentTime;
}

function updateAdHoverTimer(event: string, slot: SlotMetrics) {
  const currentTime = timeInSeconds();
  let diff = 0;

  if (slot.lastHoverStarted !== 0) {
    diff = currentTime - slot.lastHoverStarted;
  }

  if (event === "mouseleave") {
    slot.totalHoverTime += diff;
  } else {
    slot.lastHoverStarted = currentTime;
  }
}

function adListeners(target: HTMLElement, slot: SlotMetrics): void {
  target.addEventListener("mouseenter", () => {
    updateAdHoverTimer("mouseenter", slot);
  });

  target.addEventListener("mouseleave", () => {
    updateAdHoverTimer("mouseleave", slot);
  });
}
