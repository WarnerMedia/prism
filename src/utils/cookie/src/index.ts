import findTopDomain from "@segment/top-domain";
import { cookieHelper, debug } from "./helper";

export interface CookieModule {
  options(options?: CookieOptions): void | CookieOptions;
  set(name: string, value: string | object | boolean): boolean;
  get(name: string): string | object | boolean | null;
  remove(name: string): boolean;
}

export interface CookieOptions {
  maxage?: number;
  expires?: Date;
  domain?: string | null;
  path?: string;
  secure?: boolean;
  samesite?: string;
  useRootDomain?: boolean;
  encode?: boolean;
}

export class Cookie implements CookieModule {
  private _options: CookieOptions = {};

  constructor(options?: CookieOptions) {
    this.options = this.options.bind(this);
    this.set = this.set.bind(this);
    this.get = this.get.bind(this);
    this.getAll = this.getAll.bind(this);
    this.remove = this.remove.bind(this);
    this.options(options);
  }

  options(options?: CookieOptions): undefined | CookieOptions {
    if (arguments.length === 0) return this._options;

    options = options || {};

    let domain: CookieOptions["domain"] =
      options.domain || `.${findTopDomain(window.location.href)}`;

    if (domain === ".") {
      domain = null;
    }

    this._options = {
      maxage: 31536000000,
      path: "/",
      samesite: "Lax",
      encode: true,
      ...options,
      domain,
    };

    // http://curl.haxx.se/rfc/cookie_spec.html
    // https://publicsuffix.org/list/effective_tld_names.dat
    //
    // try setting a dummy cookie with the options
    // if the cookie isn't set, it probably means
    // that the domain is on the public suffix list
    // like myapp.deployment.com or localhost / ip.
    this.set("psm:test", true, this._options);
    if (!this.get("psm:test")) {
      debug(
        "[Cookie]: psm:test cookie could not be set at domain: ",
        this._options.domain
      );
      const reg = new RegExp(document.location.hostname, "g");
      if (document.referrer !== "" && document.referrer.search(reg) === -1) {
        this._options.samesite = "None";
      } else {
        this._options.domain = null;
      }
    }
    this.remove("psm:test");

    debug("[COOKIE]: this._options", this._options);
  }

  set(
    name: string,
    value: string | object | boolean,
    options: Partial<CookieOptions> = {}
  ): boolean {
    const opts = { ...this._options, ...options };

    debug("[COOKIE]: this.set() options", opts);

    try {
      if (typeof value !== "string") {
        value = JSON.stringify(value);
      }
      // @ts-ignore
      cookieHelper(name, value === "null" ? null : value, opts);
      return true;
    } catch (err) {
      debug("[COOKIE]: this.set() encountered an error", err);
      return false;
    }
  }

  get(name: string): string | object | boolean | null {
    const value = cookieHelper(name) as string | undefined;

    try {
      return JSON.parse(value);
    } catch (err) {
      // debug('[COOKIE]: this.get() encountered an error', err);
      return value || null;
    }
  }

  getAll() {
    return cookieHelper();
  }

  remove(name: string): boolean {
    try {
      cookieHelper(name, null, { ...this._options });
      return true;
    } catch (err) {
      debug("[COOKIE]: this.remove() encountered an error", err);
      return false;
    }
  }
}

export const cookie = new Cookie();
export { findTopDomain };
