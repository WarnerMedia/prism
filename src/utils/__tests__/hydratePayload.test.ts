/* eslint-disable @typescript-eslint/no-explicit-any */
import { v4 as uuid } from "uuid";
import { cookie } from "../cookie/src";
import {
  getAdsGUID,
  getEcid,
  getKruxId,
  getConvivaId,
  getOptanonConsentCookie,
  getMetadataPropValue,
  getContentMetadata,
  hydratePayload,
} from "../hydratePayload";
import { cookiesEnabled, localStorageAccessible } from "../browser";
import { payloadCore } from "../../payloadCore";
import { Psm } from "../../Psm";
import { clearMockIntl, mockIntl, clearAllCookies } from "../testUtils";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "00000000-0000-0000-0000-000000000000"),
}));

jest.mock("../browser", () => ({
  ...(jest.requireActual("../browser") as object),
  cookiesEnabled: jest.fn().mockReturnValue(true),
  localStorageAccessible: jest.fn().mockReturnValue(true),
}));

jest.mock("../hydratePayload", () => ({
  ...(jest.requireActual("../hydratePayload") as object),
}));

jest.mock("../../../package.json", () => ({
  version: "0.0.0",
}));

const psmConfig = {
  brand: "Prism",
  subBrand: "Prism-Test",
  cookieDomain: ".psm.test",
  psmEnvironment: "TEST",
};
const timestamp = "2020-12-01T00:00:00.000Z";
const cookiesEnabledMock = cookiesEnabled as jest.MockedFunction<
  typeof cookiesEnabled
>;
const localStorageAccessibleMock = localStorageAccessible as jest.MockedFunction<
  typeof localStorageAccessible
>;
const win = window as any;

describe("hydratePayload()", () => {
  beforeEach(() => {
    mockIntl();
  });

  afterEach(() => {
    localStorage.clear();
    clearAllCookies();
    clearMockIntl();
  });

  it("should add all expected fields to a payload", () => {
    // set up Psm instance and core to pass to hydratePayload
    const psm = new Psm();
    const core = payloadCore((data) => {
      // once the payload is built and we call a "track" method from the core
      // we can visually check the payload in snapshots
      const payload = data.build();
      expect(payload).toMatchSnapshot();
    });

    // stub out Psm class properties
    psm.ukid = uuid();
    psm.csid = uuid();
    psm.config = psmConfig;

    // populate payload with values
    hydratePayload(psm, core);
    // call a "track" method to build the payload
    core.trackIdentityRegistration("identity test", timestamp);
  });

  describe("getKruxId()", () => {
    it("should get a Krux ID from localStorage", () => {
      localStorage.setItem("kxkuid", "foo");

      expect(getKruxId()).toBe("foo");
    });

    it("should get a Krux ID from the kxkuid cookie", () => {
      localStorageAccessibleMock.mockReturnValueOnce(false);
      cookie.set("kxkuid", "foo");

      expect(getKruxId()).toBe("foo");
    });

    it("should return null when localStorage and cookies are not available", () => {
      cookiesEnabledMock.mockReturnValueOnce(false);
      localStorageAccessibleMock.mockReturnValueOnce(false);
      localStorage.setItem("kxkuid", "foo");
      cookie.set("kxkuid", "foo");

      expect(getKruxId()).toBeNull();
    });
  });

  describe("getEcid()", () => {
    it("should return the ecid when s_ecid has a value", () => {
      cookie.set("s_ecid", '"11111"');
      expect(getEcid()).toBe("11111");
    });

    it("should return the ecid when s_ecid does not have a value but s_vi does", () => {
      cookie.set("s_vi", '"00000"');
      expect(getEcid()).toBe("00000");
    });

    it("should return the AMCV cookie ECID value when s_ecid does not have a value and s_vi does not either", () => {
      cookie.set(
        "AMCV_000000000000000000000000@ExampleOrg",
        '"MCMID|22222|MCAAMLH-1234567890"'
      );
      expect(getEcid()).toBe("22222");
    });

    it("should return null when s_ecid & s_vi & amcv have no value", () => {
      expect(getEcid()).toBeNull();
    });
  });

  describe("getConvivaId()", () => {
    it("should get Conviva/sdkConfig from localStorage", () => {
      localStorage.setItem(
        "Conviva/sdkConfig",
        '{ "clId": "123456.123456.123456.123456" }'
      );

      expect(getConvivaId()).toBe("123456.123456.123456.123456");
    });

    it("should get Conviva.sdkConfig from localStorage", () => {
      localStorage.setItem(
        "Conviva.sdkConfig",
        '{ "clId": "123456.123456.123456.1234567" }'
      );

      expect(getConvivaId()).toBe("123456.123456.123456.1234567");
    });

    it("should return null when localStorage is not available", () => {
      localStorageAccessibleMock.mockReturnValueOnce(false);
      localStorage.setItem(
        "Conviva.sdkConfig",
        '{ "clId": "123456.123456.123456.123456" }'
      );

      expect(getConvivaId()).toBeNull();
    });

    it("should return null when the value in Conviva/sdkConfig is not JSON", () => {
      localStorage.setItem("Conviva/sdkConfig", "Im a cat");

      expect(getConvivaId()).toBeNull();
    });
  });

  describe("getOptanonConsentCookie()", () => {
    it("should return optanonConsent cookie when it has a value", () => {
      cookie.set(
        "OptanonConsent",
        "isIABGlobal=false&datestamp=Tue+Jun+15+2021+17%3A12%3A45+GMT-0500+(Central+Daylight+Time)&version=6.10.0&hosts=&landingPath=NotLandingPage&groups=xx%3A1&AwaitingReconsent=false&geolocation=US%3BIL"
      );
      expect(getOptanonConsentCookie()).toBe(
        "isIABGlobal=false&datestamp=Tue+Jun+15+2021+17%3A12%3A45+GMT-0500+(Central+Daylight+Time)&version=6.10.0&hosts=&landingPath=NotLandingPage&groups=xx%3A1&AwaitingReconsent=false&geolocation=US%3BIL"
      );
    });

    it("should return null if cookies are diabled", () => {
      cookiesEnabledMock.mockReturnValueOnce(false);
      cookie.set(
        "OptanonConsent",
        "isIABGlobal=false&datestamp=Tue+Jun+15+2021+17%3A12%3A45+GMT-0500+(Central+Daylight+Time)&version=6.10.0&hosts=&landingPath=NotLandingPage&groups=xx%3A1&AwaitingReconsent=false&geolocation=US%3BIL"
      );

      expect(getOptanonConsentCookie()).toBeNull();
    });

    it("should return null if cookie does not exist", () => {
      expect(getOptanonConsentCookie()).toBeNull();
    });
  });

  describe("getAdsGUID()", () => {
    it("should retrieve the id from the ug cookie", () => {
      cookie.set("ug", "test-guid");

      expect(getAdsGUID()).toBe("test-guid");
    });

    it("should return null if cookies are diabled", () => {
      cookiesEnabledMock.mockReturnValueOnce(false);

      expect(getAdsGUID()).toBeNull();
    });
  });
});
