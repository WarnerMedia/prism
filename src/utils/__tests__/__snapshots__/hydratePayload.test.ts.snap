// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hydratePayload() should add all expected fields to a payload 1`] = `
Object {
  "adsProperties": Object {
    "ads": Array [],
  },
  "brand": "Prism",
  "companyName": "Prism",
  "consentProperties": Object {
    "consentRule": "Other",
  },
  "contentMetadata": Object {
    "page": Object {},
    "video": Object {},
  },
  "device": {"type":"","totalWidth":0,"totalHeight":0,"screenResolution":"0x0","viewportSize":"1024x768"},
  "eventId": "00000000-0000-0000-0000-000000000000",
  "eventName": "identity test",
  "eventProperties": Object {
    "automatedTest": false,
    "cookies": Object {},
    "cookiesEnabled": true,
    "doNotTrack": false,
    "featureFlagValues": Object {},
    "heartbeat": Object {
      "currentScrollPosition": 0,
      "didScrollUpDuringInterval": false,
      "engagedTime": 0,
      "interval": 30,
      "maxScrollDepth": 0,
      "totalTime": 0,
    },
    "optimizelyFlagValues": Object {
      "optimizelyInBrainEnabled": false,
      "optimizelyInControlOfInBrain": false,
    },
  },
  "eventTimestamp": "2020-12-01T00:00:00.000Z",
  "eventType": "identity",
  "iabConsentCategories": Object {},
  "ids": Object {
    "csid": "Unknown",
  },
  "library": Object {
    "initConfig": Object {
      "brand": "Prism",
      "cookieDomain": ".psm.test",
      "psmEnvironment": "TEST",
      "subBrand": "Prism-Test",
    },
    "name": "PrismJS",
    "version": "0.0.0",
  },
  "location": Object {
    "country": "",
    "language": "en",
    "locale": "en-US",
    "state": "",
    "timezone": "UTC",
  },
  "navigationProperties": Object {
    "path": "/",
    "referrer": "",
    "rootDomain": "",
    "search": "",
    "title": "",
    "url": "http://localhost/",
  },
  "subBrand": "Prism-Test",
  "ukid": "00000000-0000-0000-0000-000000000000",
}
`;
